core/unix/src/TUnixSystem.cxx

```diff
@@ -568,6 +573,94 @@ TUnixSystem::~TUnixSystem()
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Initialize Unix system interface.
+extern "C"
+{
+   static int FullWrite(int fd, const char *text)
+   {
+      const char *buffer = text;
+      size_t count = strlen(text);
+      ssize_t written = 0;
+      while (count) {
+         written = write(fd, buffer, count);
+         if (written == -1) {
+          if (errno == EINTR) { continue; }
+          else { return -errno; }
+         }
+         count -= written;
+         buffer += written;
+      }
+      return 0;
+   }
+
+   static int FullRead(int fd, char *inbuf, size_t len, int timeout=-1) {
+      char *buf = inbuf;
+      size_t count = len;
+      ssize_t complete = 0;
+      std::chrono::time_point<std::chrono::steady_clock> endTime = std::chrono::steady_clock::now() + std::chrono::seconds(timeout);
+      int flags;
+      if (timeout < 0) {
+         flags = O_NONBLOCK;  // Prevents us from trying to set / restore flags later.
+      } else if ((-1 == (flags = fcntl(fd, F_GETFL)))) {
+         return -errno;
+      } else { }
+      if ((flags & O_NONBLOCK) != O_NONBLOCK) {
+         if (-1 == fcntl(fd, F_SETFL, flags | O_NONBLOCK)) {
+            return -errno;
+         }
+      }
+      while (count) {
+         if (timeout >= 0) {
+            struct pollfd pollInfo{fd, POLLIN, 0};
+            int msRemaining = std::chrono::duration_cast<std::chrono::milliseconds>(endTime-std::chrono::steady_clock::now()).count();
+            if (msRemaining > 0) {
+               if (poll(&pollInfo, 1, msRemaining) == 0) {
+                  if ((flags & O_NONBLOCK) != O_NONBLOCK) {
+                     fcntl(fd, F_SETFL, flags);
+                  }
+                  return -ETIMEDOUT;
+               }
+            } else if (msRemaining < 0) {
+               if ((flags & O_NONBLOCK) != O_NONBLOCK) {
+                  fcntl(fd, F_SETFL, flags);
+               }
+               return -ETIMEDOUT;
+            } else { }
+         }
+         complete = read(fd, buf, count);
+         if (complete == -1) {
+            if (errno == EINTR) { continue; }
+            else if ((errno == EAGAIN) || (errno == EWOULDBLOCK)) { continue; }
+            else {
+               int origErrno = errno;
+               if ((flags & O_NONBLOCK) != O_NONBLOCK) {
+                  fcntl(fd, F_SETFL, flags);
+               }
+               return -origErrno;
+            }
+         }
+         count -= complete;
+         buf += complete;
+      }
+      if ((flags & O_NONBLOCK) != O_NONBLOCK) {
+         fcntl(fd, F_SETFL, flags);
+      }
+      return 0;
+   }
+
+   static int FullErrWrite(const char *text) {
+      return FullWrite(2, text);
+   }
+
+}
+
+std::unique_ptr<std::thread> TUnixSystem::fHelperThread;
```

Comment:
Static declarations ought to be towards the beginning of the files (before any function implementation).